src/kernel/arch/x86/pit.zig

```diff
@@ -148,111 +185,135 @@ inline fn sendCommand(cmd: u8) void {
 ///                      OCW_SELECT_COUNTER_1 or OCW_SELECT_COUNTER_2 only.
+///     IN freq: u32   - The frequency that the counter operates at. Any frequency that is between
+///                      0..18 (inclusive) will default to 19Hz. Any frequency above MAX_FREQUENCY
+///                      will default to MAX_FREQUENCY.
 ///     IN mode: u8    - The mode of operation that the counter will operate in. See The modes
 ///                      definition above to chose which mode the counter is to run at.
 ///
 fn setupCounter(comptime counter: u8, freq: u32, mode: u8) void {
-    var reload_value: u32 = 0x10000; // 65536, the slowest possible frequency
+    // 65536, the slowest possible frequency. Roughly 19Hz
+    var reload_value = u32(0x10000);
 
+    // The lowest possible frequency is 18Hz.
+    // MAX_FREQUENCY / 18 > u16 N
+    // MAX_FREQUENCY / 19 < u16 Y
     if (freq > 18) {
-        // The fastest possible frequency if frequency is too high
-        reload_value = 1;
-
         if (freq < MAX_FREQUENCY) {
-            reload_value = MAX_FREQUENCY / freq;
-            var rem: u32 = MAX_FREQUENCY % freq;
-
-            // Is the remainder more than half
-            if (rem > MAX_FREQUENCY / 2) {
-                // Round up
-                reload_value += 1;
-            }
+            // Rounded integer division
+            reload_value = (MAX_FREQUENCY + freq - 1) / freq;
+        } else {
+            // The fastest possible frequency if frequency is too high
+            reload_value = 1;
         }
     }
 
-    // Get the u16 version as this is what will be loaded into the PIT
-    var reload_val_16: u16 = @truncate(u16, reload_value);
-
     // Update the frequency with the actual one that the PIT will be using
-    var frequency: u32 = MAX_FREQUENCY / reload_value;
-    var rem: u32 = MAX_FREQUENCY % reload_value;
+    // Rounded integer division
+    const frequency = (MAX_FREQUENCY + reload_value - 1) / reload_value;
 
-    // Is the remainder more than half
-    if (rem > MAX_FREQUENCY / 2) {
-        // Round up
-        frequency += 1;
-    }
+    // Calculate the amount of nanoseconds between interrupts
+    time_ns = u32(1000000000) / frequency;
 
-    // Calculate the amount of time between IRQs
-    time_ms = reload_value * u32(1000);
-    time_under_1_ms = time_ms % MAX_FREQUENCY;
-    time_ms /= MAX_FREQUENCY;
+    // Calculate the number of picoseconds, the left over from nanoseconds
+    time_under_1_ns = ((u32(1000000000) % frequency) * u32(1000)) / frequency;
 
-    comptime const port: u16 = switch (counter) {
+    // Select the port to set up
+    comptime const port = switch (counter) {
         OCW_SELECT_COUNTER_0 => COUNTER_0_REGISTER,
         OCW_SELECT_COUNTER_1 => COUNTER_1_REGISTER,
         OCW_SELECT_COUNTER_2 => COUNTER_2_REGISTER,
-        else => unreachable,
+        else => @compileError("Invalid counter"),
```

Comment:
Using an enum would remove the need for a compile error and for `counter` to be comptime, but I know you don't like enums. The enum could even have a function that returns its register.

Reply:
Ill have ago and see how it looks